core/drivers/crypto/hisilicon/hpre_ecc.c

```diff
@@ -0,0 +1,997 @@
+				       ecc_dh->x_bytes);
+	if (ret) {
+		EMSG("Fail to transfer ecdh gx from crypto_bin to hpre_bin");
+		return ret;
+	}
+
+	ret = hpre_bin_from_crypto_bin(y, y, msg->key_bytes,
+				       ecc_dh->y_bytes);
+	if (ret)
+		EMSG("Fail to transfer ecdh gy from crypto_bin to hpre_bin");
+
+	return ret;
+}
+
+static enum hisi_drv_status
+hpre_ecc_dh_params_fill(struct hpre_ecc_curve *curve, struct hpre_ecc_msg *msg,
+			struct bignum *privkey, struct ecc_public_key *pubkey)
+{
+	struct hpre_ecc_dh *ecc_dh = &msg->param_size.ecc_dh;
+	uint8_t *p = msg->key;
+	uint8_t *a = p + msg->key_bytes;
+	uint8_t *d = a + msg->key_bytes;
+	uint8_t *b = d + msg->key_bytes;
+	uint8_t *x = msg->in;
+	uint8_t *y = x + msg->key_bytes;
+	enum hisi_drv_status ret = HISI_QM_DRVCRYPT_NO_ERR;
+
+	memcpy(p, curve->p, msg->curve_bytes);
+	memcpy(a, curve->a, msg->curve_bytes);
+	crypto_bignum_bn2bin(privkey, d);
+	ecc_dh->d_bytes = crypto_bignum_num_bytes(privkey);
+	if (is_all_zero(d, ecc_dh->d_bytes, "ecc dh d"))
+		return HISI_QM_DRVCRYPT_EINVAL;
+
+	memcpy(b, curve->b, msg->curve_bytes);
+
+	ecc_dh->x_bytes = msg->curve_bytes;
+	ecc_dh->y_bytes = msg->curve_bytes;
+	if (!pubkey) { /* gen key pair */
+		memcpy(x, curve->x, ecc_dh->x_bytes);
+		memcpy(y, curve->y, ecc_dh->y_bytes);
+	} else { /* do shared secret */
+		crypto_bignum_bn2bin(pubkey->x, x);
+		ecc_dh->x_bytes = crypto_bignum_num_bytes(pubkey->x);
+		crypto_bignum_bn2bin(pubkey->y, y);
+		ecc_dh->y_bytes = crypto_bignum_num_bytes(pubkey->y);
+	}
+
+	ret = hpre_ecc_dh_transfer_key(msg);
+	if (ret)
+		return ret;
+
+	ret = hpre_ecc_dh_transfer_in(msg);
+
+	return ret;
+}
+
+static enum hisi_drv_status hpre_ecc_dh_params_alloc(struct hpre_ecc_msg *msg)
+{
+	uint32_t size = HPRE_ECC_DH_TOTAL_BUF_SIZE(msg->key_bytes);
+	uint8_t *data = NULL;
+
+	data = calloc(1, size);
+	if (!data) {
+		EMSG("Fail to alloc ecc dh total buf");
+		return HISI_QM_DRVCRYPT_ENOMEM;
+	}
+
+	msg->key = data;
+	msg->key_dma = virt_to_phys(msg->key);
+
+	msg->in = msg->key + ECC_DH_KEY_SIZE(msg->key_bytes);
+	msg->in_dma = msg->key_dma + ECC_DH_KEY_SIZE(msg->key_bytes);
+	msg->out = msg->in + ECC_DH_IN_SIZE(msg->key_bytes);
+	msg->out_dma = msg->in_dma + ECC_DH_IN_SIZE(msg->key_bytes);
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static void hpre_ecc_request_deinit(struct hpre_ecc_msg *msg)
+{
+	if (msg->key) {
+		memzero_explicit(msg->key, ECC_DH_KEY_SIZE(msg->key_bytes));
+		free(msg->key);
+		msg->key = NULL;
+	}
+}
+
+static TEE_Result hpre_ecc_request_init(struct hpre_ecc_curve *curve,
+					struct hpre_ecc_msg *msg,
+					struct bignum *d,
+					struct ecc_public_key *pubkey)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	msg->alg_type = HPRE_ALG_ECDH_MULTIPLY;
+
+	if (curve->id == TEE_ECC_CURVE_SM2)
+		msg->sm2_sp = true;
+
+	msg->curve_bytes = BITS_TO_BYTES(curve->key_bits);
+	msg->key_bytes = hpre_ecc_get_hw_kbytes(curve->key_bits);
+	if (!msg->key_bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = hpre_ecc_dh_params_alloc(msg);
```

Comment:
Type mismatch here, hpre_ecc_dh_params_alloc returns of type enum hisi_drv_status which is assigned to type TEE_Result.
and i see enum hisi_drv_status declaration is missing in this patch.

Reply:
solved. enum hisi_drv_status declaration was merged in patch about qm.